libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1347,19 +1397,76 @@ string YulUtilFunctions::updateStorageValueFunction(Type const& _type, std::opti
 			("functionName", functionName)
 			("update",
 				_offset.has_value() ?
-					updateByteSliceFunction(_type.storageBytes(), *_offset) :
-					updateByteSliceFunctionDynamic(_type.storageBytes())
+					updateByteSliceFunction(_toType.storageBytes(), *_offset) :
+					updateByteSliceFunctionDynamic(_toType.storageBytes())
 			)
 			("offset", _offset.has_value() ? "" : "offset, ")
-			("prepare", prepareStoreFunction(_type))
+			("prepare", prepareStoreFunction(_toType))
 			.render();
 		}
 		else
 		{
-			if (_type.category() == Type::Category::Array)
-				solUnimplementedAssert(false, "");
-			else if (_type.category() == Type::Category::Struct)
+			if (_toType.category() == Type::Category::Array)
 				solUnimplementedAssert(false, "");
+			else if (_toType.category() == Type::Category::Struct)
+			{
+				solAssert(_fromType, "");
+				solAssert(_fromType->category() == Type::Category::Struct, "");
+				auto const& fromStructType = dynamic_cast<StructType const&>(*_fromType);
+				auto const& toStructType = dynamic_cast<StructType const&>(_toType);
+				solAssert(fromStructType.structDefinition() == toStructType.structDefinition(), "");
+				solAssert(fromStructType.location() != DataLocation::Storage, "");
+				solUnimplementedAssert(_offset.has_value() && _offset.value() == 0, "");
+
+				Whiskers templ(R"(
+					function <functionName>(slot, value) {
+						<?fromCalldata>
+							let valueMem := <convertToMemory>(value)
+						<!fromCalldata>
+							let valueMem := value
+						</fromCalldata>
+						<#member>
+							<updateMemberCall>
+						</member>
+					}
+				)");
+				templ("functionName", functionName);
+				templ("fromCalldata", fromStructType.location() == DataLocation::CallData);
+				if (fromStructType.location() == DataLocation::CallData)
+					templ("convertToMemory", conversionFunction(
+						fromStructType,
+						*TypeProvider::structType(toStructType.structDefinition(), DataLocation::Memory)
+					));
+
+				MemberList::MemberMap toStructMembers = toStructType.nativeMembers(nullptr);
+				MemberList::MemberMap fromStructMembers = fromStructType.nativeMembers(nullptr);
+
+				vector<map<string, string>> memberParams(toStructMembers.size());
+				for (size_t i = 0; i < toStructMembers.size(); ++i)
+				{
+					solAssert(toStructMembers[i].type->memoryHeadSize() == 32, "");
+					bool isStruct = toStructMembers[i].type->category() == Type::Category::Struct;
+					auto const& [slotDiff, offset] = toStructType.storageOffsetsOfMember(toStructMembers[i].name);
+					memberParams[i]["updateMemberCall"] = Whiskers(R"(
+						<updateMember>(add(slot, <memberStorageSlotDiff>), <?notStruct><memberStorageOffset>,</notStruct> <loadFromMemory>(add(valueMem, <memberMemoryOffset>)))
+					)")
+					("notStruct", !isStruct)
```

Comment:
```suggestion
					("hasOffset", !isStruct)
```

Reply:
What about arrays?

Reply:
I think there is no support for arrays here. I planned to add that in separate PR, to not over-complicate this one.